exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +319,264 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+
+			patchHelper, err := patch.NewHelper(infraMachine, r.Client)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create patch helper for %s", klog.KObj(infraMachine))
+			}
+
+			ownerRefs = util.EnsureOwnerRef(ownerRefs, *machineRef)
+			infraMachine.SetOwnerReferences(ownerRefs)
+
+			if err := patchHelper.Patch(ctx, infraMachine); err != nil {
+				return errors.Wrapf(err, "failed to patch %s", klog.KObj(infraMachine))
+			}
+		}
+	}
+
+	return nil
+}
+
+// getNewMachine creates a new Machine object.
+func getNewMachine(mp *expv1.MachinePool, infraMachine *unstructured.Unstructured) *clusterv1.Machine {
+	infraRef := corev1.ObjectReference{
+		APIVersion: infraMachine.GetAPIVersion(),
+		Kind:       infraMachine.GetKind(),
+		Name:       infraMachine.GetName(),
+		Namespace:  infraMachine.GetNamespace(),
+	}
+
+	machine := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-", mp.Name)),
+			// Note: by setting the ownerRef on creation we signal to the Machine controller that this is not a stand-alone Machine.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(mp, mp.GroupVersionKind())},
+			Namespace:       mp.Namespace,
+			Labels:          make(map[string]string),
+			Annotations:     make(map[string]string),
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName:       mp.Spec.ClusterName,
+			InfrastructureRef: infraRef,
+		},
+	}
+
+	for k, v := range mp.Spec.Template.Annotations {
+		machine.Annotations[k] = v
+	}
+
+	// Set the labels from machinePool.Spec.Template.Labels as labels for the new Machine.
+	// Note: We can't just set `machinePool.Spec.Template.Labels` directly and thus "share" the labels
+	// map between Machine and machinePool.Spec.Template.Labels. This would mean that adding the
+	// MachinePoolNameLabel later on the Machine would also add the labels to machinePool.Spec.Template.Labels
+	// and thus modify the labels of the MachinePool.
+	for k, v := range mp.Spec.Template.Labels {
+		machine.Labels[k] = v
+	}
+
+	// Enforce that the MachinePoolNameLabel and ClusterNameLabel are present on the Machine.
+	machine.Labels[clusterv1.MachinePoolNameLabel] = capilabels.MustFormatValue(mp.Name)
+	machine.Labels[clusterv1.ClusterNameLabel] = mp.Spec.ClusterName
+
+	return machine
+}
+
+// infraMachineToMachinePoolMapper is a mapper function that maps an InfraMachine to the MachinePool that owns it.
+// This is used to trigger an update of the MachinePool when a InfraMachine is changed.
+func (r *MachinePoolReconciler) infraMachineToMachinePoolMapper(ctx context.Context, o client.Object) []ctrl.Request {
+	log := ctrl.LoggerFrom(ctx)
+
+	machinePoolList := &expv1.MachinePoolList{}
+	labels := o.GetLabels()
+	selector := map[string]string{}
+	clusterName, ok := labels[clusterv1.ClusterNameLabel]
+	if !ok {
+		selector = map[string]string{clusterv1.ClusterNameLabel: clusterName}
+	}
```

Comment:
```suggestion
	selector := map[string]string{}
	if clusterName, ok := labels[clusterv1.ClusterNameLabel]; ok {
		selector = map[string]string{clusterv1.ClusterNameLabel: clusterName}
	}
```
(the `!ok` is wrong)